File: nuxt_side/components/Kiwi/models/search_preset_info.js
import MemoryRecord from 'js-memory-record'

export class SearchPresetInfo extends MemoryRecord {
  static get define() {
    return [
      { key: "新着",     },
      { key: "視聴数",   },
      { key: "居飛車",   },
      { key: "振り飛車", },
      { key: "右玉",     },
      { key: "履歴",     },
      { key: "公開",     },
      { key: "限定公開", },
      { key: "非公開",   },
    ]
  }
}

File: app/models/kiwi/search_preset_info.rb
module Kiwi
  class SearchPresetInfo
    include ApplicationMemoryRecord
    memory_record [
      { key: "新着",     func: -> s, params { s.public_only_with_user(params)                                                            }, },
      { key: "視聴数",   func: -> s, params { s.public_only_with_user(params).order(access_logs_count: :desc)                                    }, },
      { key: "居飛車",   func: -> s, params { s.public_only_with_user(params).tagged_with("居飛車")                                      }, },
      { key: "振り飛車", func: -> s, params { s.public_only_with_user(params).tagged_with("振り飛車")                                    }, },
      { key: "右玉",     func: -> s, params { s.public_only_with_user(params).tagged_with(::Swars::UserInfo.migigyoku_family, any: true) }, },
      { key: "履歴",     func: -> s, params { params[:current_user] ? params[:current_user].kiwi_access_logs.uniq_histories : s.none         }, },
      { key: "公開",     func: -> s, params { params[:current_user] ? params[:current_user].kiwi_bananas.folder_eq(:public) : s.none               }, },
      { key: "限定公開", func: -> s, params { params[:current_user] ? params[:current_user].kiwi_bananas.folder_eq(:limited) : s.none              }, },
      { key: "非公開",   func: -> s, params { params[:current_user] ? params[:current_user].kiwi_bananas.folder_eq(:private) : s.none               }, },
    ]
  end
end
